File: java/decouple-monolith/solution/src/main/java/compliance/temporal/ComplianceWorkerApp.java
package compliance.temporal;

import compliance.ComplianceChecker;
import compliance.temporal.activity.ComplianceActivityImpl;
import compliance.temporal.workflow.ComplianceWorkflowImpl;
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowClientOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;
import io.temporal.worker.Worker;
import io.temporal.worker.WorkerFactory;

/**
 * Compliance team's worker — handles Nexus requests from Payments.
 * Task queue: "compliance-risk"
 *
 * Registers three things:
 *   1. ComplianceWorkflowImpl — the workflow that wraps the activity
 *   2. ComplianceActivityImpl — the activity that runs the checker
 *   3. ComplianceNexusServiceImpl — the Nexus handler that launches the workflow
 */
public class ComplianceWorkerApp {

    public static void main(String[] args) {
        // C — Connect to Temporal (compliance-namespace)
        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
        WorkflowClientOptions clientOptions = WorkflowClientOptions.newBuilder()
                .setNamespace("compliance-namespace")
                .build();
        WorkflowClient client = WorkflowClient.newInstance(service, clientOptions);

        // R — Create factory and worker
        WorkerFactory factory = WorkerFactory.newInstance(client);
        Worker worker = factory.newWorker("compliance-risk");

        // W — Wire workflow, activity, and Nexus handler
        worker.registerWorkflowImplementationTypes(ComplianceWorkflowImpl.class);
        worker.registerActivitiesImplementations(new ComplianceActivityImpl(new ComplianceChecker()));
        worker.registerNexusServiceImplementation(new ComplianceNexusServiceImpl());

        // L — Launch
        factory.start();

        System.out.println("=========================================================");
        System.out.println("  Compliance Worker started on: compliance-risk");
        System.out.println("  Namespace: compliance-namespace");
        System.out.println("  Registered: ComplianceWorkflow, ComplianceActivity,");
        System.out.println("              ComplianceNexusServiceImpl (async + sync handlers)");
        System.out.println("  Waiting for Nexus requests from Payments team...");
        System.out.println("=========================================================");
    }
}

File: kotlin/decouple-monolith/solution/src/main/kotlin/compliance/temporal/ComplianceWorkerApp.kt
package compliance.temporal

import compliance.ComplianceChecker
import compliance.temporal.activity.ComplianceActivityImpl
import compliance.temporal.workflow.ComplianceWorkflowImpl
import io.temporal.client.WorkflowClient
import io.temporal.client.WorkflowClientOptions
import io.temporal.serviceclient.WorkflowServiceStubs
import io.temporal.worker.WorkerFactory

/**
 * The Compliance team's Worker. Handles Nexus requests from Payments.
 *
 * The Task Queue MUST match --target-task-queue on the Nexus Endpoint.
 */
fun main() {
    // Client (newInstance): connect, scoped to the Compliance team's own Namespace.
    val service = WorkflowServiceStubs.newLocalServiceStubs()
    val client = WorkflowClient.newInstance(
        service,
        WorkflowClientOptions.newBuilder()
            .setNamespace("compliance-namespace")
            .build(),
    )

    // Worker (newWorker): poll the Task Queue the Endpoint points at.
    val factory = WorkerFactory.newInstance(client)
    val taskQueue = "compliance-risk"
    val worker = factory.newWorker(taskQueue)

    // Register (register*): everything this Worker is responsible for. The Nexus handler
    // is what makes this team callable from Payments. Without it the Worker still starts,
    // it just never answers a Nexus Operation.
    worker.registerWorkflowImplementationTypes(ComplianceWorkflowImpl::class.java)
    worker.registerActivitiesImplementations(ComplianceActivityImpl(ComplianceChecker()))
    worker.registerNexusServiceImplementation(ComplianceNexusServiceImpl())

    // Start (start): begin polling.
    factory.start()

    println("=========================================================")
    println("  Compliance Worker started on: $taskQueue")
    println("  Namespace: compliance-namespace")
    println("  Registered: ComplianceWorkflow, ComplianceActivity, ComplianceNexusService")
    println("=========================================================")
}
